src/assign.c

```diff
@@ -153,7 +153,7 @@ static SEXP shallow(SEXP dt, SEXP cols, R_len_t n)
   SEXP newdt = PROTECT(allocVector(VECSXP, n)); protecti++;   // to do, use growVector here?
   SET_ATTRIB(newdt, shallow_duplicate(ATTRIB(dt)));
   SET_OBJECT(newdt, OBJECT(dt));
```

Comment:
Backport `SHALLOW_DUPLICATE_ATTRIB` and use it here?

https://github.com/wch/r-source/blob/8e4314e5929e2a2f890fda48301324865b79d2c9/src/main/memory.c#L3944-L3948

Reply:
Nice find... I think we can just go ahead and use it?

![image](https://github.com/user-attachments/assets/8dd7953a-2225-4cc5-9ac7-5523315c06b4)

I would leave this as a TODO for now, given the experience with `isDataFrame()`/ #6235 -- let's wait for the `NOTE` to start throwing at least.

Reply:
Per the comment just below, it's available from 3.3.0 (which we now use), so I went ahead and made the switch.